src/rdb.c

```diff
@@ -2291,111 +2293,78 @@ robj *rdbLoadObject(int rdbtype, rio *rdb, sds key, redisDb* db, int rdbflags,
-            if ((expireAt != 0) && (expireAt < minExpire)) minExpire = expireAt;
+            if ((expireAt != 0) && (expireAt < minExpiredField)) minExpiredField = expireAt;
 
             /* store the values read - either to listpack or dict */
             if (o->encoding == OBJ_ENCODING_LISTPACK_EX) {
                 /* integrity - check for key duplication (if required) */
                 if (dupSearchDict) {
-                    sds field_dup = sdsnewlen(field, sdslen(field));
+                    sds field_dup = sdsnewlen(field, hfieldlen(field));
 
                     if (dictAdd(dupSearchDict, field_dup, NULL) != DICT_OK) {
                         rdbReportCorruptRDB("Hash with dup elements");
                         dictRelease(dupSearchDict);
                         decrRefCount(o);
                         sdsfree(field_dup);
                         sdsfree(value);
-                        sdsfree(field);
+                        hfieldFree(field);
                         return NULL;
                     }
                 }
 
                 /* check if the values can be saved to listpack (or should convert to dict encoding) */
-                if (sdslen(field) > server.hash_max_listpack_value ||
+                if (hfieldlen(field) > server.hash_max_listpack_value ||
                     sdslen(value) > server.hash_max_listpack_value ||
-                    !lpSafeToAdd(((listpackEx*)o->ptr)->lp, sdslen(field) + sdslen(value) + lpEntrySizeInteger(expireAt)))
+                    !lpSafeToAdd(((listpackEx*)o->ptr)->lp, hfieldlen(field) + sdslen(value) + lpEntrySizeInteger(expireAt)))
                 {
                     /* convert to hash */
-                    hashTypeConvert(o, OBJ_ENCODING_HT, hexpires);
+                    hashTypeConvert(o, OBJ_ENCODING_HT, NULL);
 
                     if (len > DICT_HT_INITIAL_SIZE) { /* TODO: this is NOT the original len, but this is also the case for simple hash, is this a bug? */
                         if (dictTryExpand(o->ptr, len) != DICT_OK) {
                             rdbReportCorruptRDB("OOM in dictTryExpand %llu", (unsigned long long)len);
                             decrRefCount(o);
                             if (dupSearchDict != NULL) dictRelease(dupSearchDict);
                             sdsfree(value);
-                            sdsfree(field);
+                            hfieldFree(field);
                             return NULL;
                         }
                     }
 
                     /* don't add the values to the new hash: the next if will catch and the values will be added there */
                 } else {
-                    listpackExAddNew(o, field, value, expireAt);
-                    sdsfree(field);
+                    listpackExAddNew(o, field, hfieldlen(field),
+                                     value, sdslen(value), expireAt);
+                    hfieldFree(field);
```

Comment:
So we used sds functions for hfield before and sanitizer did not warn us? Maybe we need some test cases that would fail on such a mistake.

Reply:
@tezc ASAN reported, but not here.
i think sds and hfield are too similar that in some cases they are the same.

Reply:
Sure I know. Still we can try to add a test where they are not similar. Any mismatch can cause a sanitizer error/test failure. e.g. we can use longer strings